lib/ui/text.dart

```diff
@@ -770,6 +770,7 @@ Int32List _encodeParagraphStyle(
   StrutStyle strutStyle,
   String ellipsis,
   Locale locale,
+  bool ellipsizedByCharacter,
```

Comment:
Can this replace the existing ellipsizing implementation?

If this yields better results, then I'd prefer to use this everywhere instead of supporting two ellipsizing modes in the text API.

Reply:
+1 We shouldn't be supporting and maintaining a mode that is worse or broken.

Reply:
I personal think this is a better way to do the ellipsis, but it does have slightly different behavior, I am not sure how big the impact is if we change the original ellipsis. What do you think @Hixie

Reply:
it looks like the we all agree to change the default behavior, I will update this pr.

Reply:
If we want to change the API we must at a minimum go through the breaking change process and give people a chance to speak up about it. This could break people.

That said, I think it is valuable to support both word-based and character-based ellipsizing.

Reply:
the problem with the default behavior is that it is a some times word-based sometime character-based
given a line that to be ellipsized:
1, if directly append the ellipsis to the end of it still fits in the width constraint, it will directly append it (word-based)
2, if it cannot fit, it will truncate the sentence until it fit(character-based)

The reason why we have above two cases is the way we do sort break, if a sentence end with a long word, it will push to next line and leave the current line with plenty of empty space that will lead to (1)

I think if we want to support both, we should fix the original behavior, too, but that is still a breaking change.

Reply:
The current behavior is essentially a bug.  I'd like to keep our API as is and just change the implementation, and I'd definitely prefer to avoid making the current behavior part of our documented spec.

Reply:
I don't quite understand your description of the original behaviour, but if it is indeed buggy then we should fix it, and that should go through the breaking change process.

If we're changing the existing behaviour then I don't have a strong opinion on what behavior we implement (whether it's word-based, ligature-based, or if we provide an enum to select either). In any case, we should add tests such as the ones I described below (those tests assume we go with the ligature-based behavior).

Reply:
Sounds good to me, I will implement character-based as default as that is what people and other framework generally use.